xwakes/wit/component.py

```diff
@@ -0,0 +1,386 @@
+from __future__ import annotations
+
+from .parameters import *
+from .utils import unique_sigfigs
+
+from typing import Optional, Callable, Tuple, Union, List
+
+import numpy as np
+
+
+def mix_fine_and_rough_sampling(start: float, stop: float, rough_points: int,
+                                fine_points: int, rois: List[Tuple[float, float]]):
+    """
+    Mix a fine and rough (geometric) sampling between start and stop,
+    refined in the regions of interest rois.
+    :param start: The lowest bound of the sampling
+    :param stop: The highest bound of the sampling
+    :param rough_points: The total number of data points to be
+                         generated for the rough grid, between start
+                         and stop.
+    :param fine_points: The number of points in the fine grid of each roi.
+    :param rois: List of unique tuples with the lower and upper bound of
+                 of each region of interest.
+    :return: An array with the sampling obtained.
+    """
+    intervals = [np.linspace(max(i, start), min(f, stop), fine_points)
+                 for i, f in rois
+                 if (start <= i <= stop or start <= f <= stop)]
+    fine_sampling_rois = np.hstack(intervals) if intervals else np.array([])
+    rough_sampling = np.geomspace(start, stop, rough_points)
+
+    return unique_sigfigs(
+        np.sort(np.concatenate((fine_sampling_rois, rough_sampling))), 7)
```

Comment:
:thinking: the sorting/concatenating part is ten times slower AND doesn't give the same result...

Reply:
I thought sortednp.merge was doing exactly this. Could you explain what it is doing since you implemented this? Also then we are missing a test which I can write

Reply:
Sorry didn't see the second part of the comment. But that's the whole point: `unique_sigfigs` is so slow that whatever we do here is in the shadow.
Still I will add a test for this.

Reply:
yes and no... it still adds 10-20% more time which maybe we could avoid. let me check what we can do

I understand now what's happening: `snp.merge` assumes the arrays are sorted, which was actually not the case ( `fine_sampling_rois` could be unsorted in case of overlapping rois). So the implementation was buggy in the first place...

for the test, why not testing indeed `mix_fine_and_rough_sampling` (but it does not necessarily have to be done in this PR)

Reply:
I changed the implementation to use the `stable` algorithm which is (I think) more optimised than the std sorting algorithm when the array is "partially sorted" as it is here. Now it seems even faster than before.

Reply:
Ok perfect this is great thanks Nicolas. This is a bottleneck when discretizing a big model so it will be a nice speed-up. I added some simple tests and a comment so that this should not break again.
In the tests I keep the sort+concatenate because it is simpler and we use small arrays there.